File: src/emitter/stats/collect-nodes.test.ts
import { describe, expect, test } from "vitest";

import { DIRECTION } from "../../visual-graph/direction.js";
import type { Direction } from "../../visual-graph/direction.js";
import { NODE_KIND } from "../../visual-graph/node-kind.js";
import { SUBGRAPH_KIND } from "../../visual-graph/subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../../visual-graph/visual-element-type.js";
import type { VisualElement } from "../../visual-graph/visual-element.js";
import type { VisualNode } from "../../visual-graph/visual-node.js";
import type { VisualSubgraph } from "../../visual-graph/visual-subgraph.js";
import { collectNodes } from "./collect-nodes.js";

const node = (id: string, line = 1): VisualNode => ({
  type: VISUAL_ELEMENT_TYPE.Node,
  id,
  kind: NODE_KIND.Variable,
  name: id,
  line,
  isJsxElement: false,
  endLine: null,
  unused: false,
  declarationKind: null,
  initIsFunction: false,
});

const sg = (
  id: string,
  elements: VisualElement[],
  direction: Direction = DIRECTION.TB,
): VisualSubgraph => ({
  type: VISUAL_ELEMENT_TYPE.Subgraph,
  id,
  kind: SUBGRAPH_KIND.Function,
  line: 1,
  direction,
  elements,
  endLine: null,
  ownerNodeId: "n_owner",
  ownerName: "owner",
});

describe("collectNodes", () => {
  test("returns top-level nodes verbatim", () => {
    const out = collectNodes([node("a", 1), node("b", 2)]);
    expect(out.map((n) => n.id)).toEqual(["a", "b"]);
  });

  test("flattens one level of subgraph nesting", () => {
    const out = collectNodes([
      node("a"),
      sg("s", [node("b"), node("c")]),
      node("d"),
    ]);
    expect(out.map((n) => n.id)).toEqual(["a", "b", "c", "d"]);
  });

  test("flattens recursively across multiple levels", () => {
    const out = collectNodes([
      sg("s1", [sg("s2", [node("deep")]), node("mid")]),
      node("top"),
    ]);
    expect(out.map((n) => n.id)).toEqual(["deep", "mid", "top"]);
  });

  test("empty input → empty output", () => {
    expect(collectNodes([])).toEqual([]);
  });

  test("subgraph with no node descendants contributes nothing", () => {
    expect(collectNodes([sg("s", [sg("inner", [])])])).toEqual([]);
  });
});

File: crates/unsnarl-emitter-stats/src/collect_nodes_test.rs
//! Pins the preorder walk: top-level nodes come through verbatim,
//! subgraphs flatten their `elements` into the same stream, and
//! empty subgraphs contribute nothing.

use unsnarl_visual_graph::direction::Direction;
use unsnarl_visual_graph::visual_element::VisualElement;
use unsnarl_visual_graph::visual_node::{BindingVisualNode, VisualNode};
use unsnarl_visual_graph::visual_subgraph::OwnedVisualSubgraph;

use super::collect_nodes;

fn node(id: &str, line: u32) -> VisualElement {
    VisualElement::from(VisualNode::from(BindingVisualNode::const_binding(
        id, id, line,
    )))
}

fn sg(id: &str, elements: Vec<VisualElement>) -> VisualElement {
    VisualElement::from(unsnarl_visual_graph::visual_subgraph::VisualSubgraph::from(
        OwnedVisualSubgraph::function(
            id,
            1,
            Some("n_owner".to_string()),
            "owner",
            elements,
            Direction::TB,
        ),
    ))
}

fn ids(out: &[&VisualNode]) -> Vec<String> {
    out.iter().map(|n| n.id().to_string()).collect()
}

#[test]
fn returns_top_level_nodes_verbatim() {
    let input = vec![node("a", 1), node("b", 2)];
    assert_eq!(
        ids(&collect_nodes(&input)),
        vec!["a".to_string(), "b".to_string()]
    );
}

#[test]
fn flattens_one_level_of_subgraph_nesting() {
    let input = vec![
        node("a", 1),
        sg("s", vec![node("b", 1), node("c", 1)]),
        node("d", 1),
    ];
    assert_eq!(
        ids(&collect_nodes(&input)),
        vec![
            "a".to_string(),
            "b".to_string(),
            "c".to_string(),
            "d".to_string()
        ]
    );
}

#[test]
fn flattens_recursively_across_multiple_levels() {
    let input = vec![
        sg("s1", vec![sg("s2", vec![node("deep", 1)]), node("mid", 1)]),
        node("top", 1),
    ];
    assert_eq!(
        ids(&collect_nodes(&input)),
        vec!["deep".to_string(), "mid".to_string(), "top".to_string()]
    );
}

#[test]
fn empty_input_to_empty_output() {
    let input: Vec<VisualElement> = Vec::new();
    assert!(collect_nodes(&input).is_empty());
}

#[test]
fn subgraph_with_no_node_descendants_contributes_nothing() {
    let input = vec![sg("s", vec![sg("inner", vec![])])];
    assert!(collect_nodes(&input).is_empty());
}
